### ISSUE TITLE
Consistency of type usage

### ISSUE THREAD

**Issue by spoenemann** [langium/langium#429]
See https://github.com/langium/langium/pull/406#discussion_r810922786: we should validate usages of `type`s for consistency. For example, this is not valid because `Foo` is not an AST node type.
```
type Foo = string;

Bar returns Foo: ...
```
The same applies to actions and extended interfaces.

**Comment by pluralia** [comment]
We validate usages of types from declared types, we should validate also types from inferred types.
For example, `Statement` infers the type `type Statement = Definition | Evaluation` and not an interface.
```
Statement:
	Definition | Evaluation;
```
So, interface `Definition` should be not able to extend `Statement` like here:
```
interface Definition extends Statement {
    name: string
    expr: Expression
    args: DeclaredParameter[]
}
```

**Comment by montymxb** [comment]
@msujew I see you're still assigned to this after it was closed and reopened. Are you still interested in working on this one, or is it okay if I go for it?

**Comment by msujew** [comment]
@montymxb Feel free to take this one 👍

**Comment by spoenemann** [comment]
I agree with that, extending an inferred type with a declared type is weird and should not be supported.

**Comment by pluralia** [comment]
I like the idea, here is some thoughts.

> it clearly separates the fact that declared types exist on a separate level from parser rules

We are going to implement [the strict mode](https://github.com/langium/langium/issues/385) specially for the case.

> it removes having to compute inferred types (and unions) to validate existing type hierarchies

We still have to do it for validation purposes.

One of the most important advantages of Langium is the brief prototyping: you can write a couple of parser rules and everything works out of the box. Sometimes you need to add declared types in addition to this pair of parser rules for minimal type shaping, in which case it's no good to prohibit extending inferred types, because you would have to add more declared types, which goes against the idea of brief prototyping "as few entities in grammar as possible".

Another point is that inferred and declared types will become semantically "unequal": declared types can be extended, while inferred types cannot.

To summarize: I think we should do it in the strict mode only and allow it by default for brief prototyping.

We shortly discussed this problem, and @montymxb suggested using warnings when inferred types are extended by declared types. I fully support this idea: we keep prototyping brief, but at the same time notify the user when a declared type is "unstable" and can be implicitly changed because of the parent inferred type.
